apis/generators/v1alpha1/generator_acr.go

```diff
@@ -0,0 +1,122 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/external-secrets/external-secrets/apis/externalsecrets/v1beta1"
+	smmeta "github.com/external-secrets/external-secrets/apis/meta/v1"
+)
+
+// ACRAccessTokenSpec defines how to generate the access token
+// e.g. how to authenticate and which registry to use.
+// see: https://github.com/Azure/acr/blob/main/docs/AAD-OAuth.md#overview
+type ACRAccessTokenSpec struct {
+	Auth ACRAuth `json:"auth"`
+	// TenantID configures the Azure Tenant to send requests to. Required for ServicePrincipal auth type.
+	TenantID string `json:"tenantId,omitempty"`
+
+	// the domain name of the ACR registry
+	// e.g. foobarexample.azurecr.io
+	ACRRegistry string `json:"registry"`
+
+	// Define the scope for the access token, e.g. pull/push access for a repository.
+	// if not provided it will return a refresh token that has full scope.
+	// Note: you need to pin it down to the repository level, there is no wildcard available.
+	//
+	// examples:
+	// repository:my-repository:pull,push
+	// repository:my-repository:pull
+	//
+	// see docs for details: https://docs.docker.com/registry/spec/auth/scope/
+	// +optional
+	Scope string `json:"scope,omitempty"`
+
+	// EnvironmentType specifies the Azure cloud environment endpoints to use for
+	// connecting and authenticating with Azure. By default it points to the public cloud AAD endpoint.
+	// The following endpoints are available, also see here: https://github.com/Azure/go-autorest/blob/main/autorest/azure/environments.go#L152
+	// PublicCloud, USGovernmentCloud, ChinaCloud, GermanCloud
+	// +kubebuilder:default=PublicCloud
+	EnvironmentType v1beta1.AzureEnvironmentType `json:"environmentType,omitempty"`
+}
+
+type ACRAuth struct {
+	// ServicePrincipal uses Azure Service Principal credentials to authenticate with Azure.
+	// +optional
+	ServicePrincipal *AzureACRServicePrincipalAuth `json:"servicePrincipal,omitempty"`
+
+	// ManagedIdentity uses Azure Managed Identity to authenticate with Azure.
+	// +optional
+	ManagedIdentity *AzureACRManagedIdentityAuth `json:"managedIdentity,omitempty"`
+
+	// WorkloadIdentity uses Azure Workload Identity to authenticate with Azure.
+	// +optional
+	WorkloadIdentity *AzureACRWorkloadIdentityAuth `json:"workloadIdentity,omitempty"`
+}
+
+type AzureACRServicePrincipalAuth struct {
+	SecretRef AzureACRServicePrincipalAuthSecretRef `json:"secretRef"`
```

Comment:
Let's continue the discussion here @gusfcarvalho @knelasevero.

> This thing surprised me a little bit when looking at the code [...] generators are equivalent to secret stores in the sense that they carry credentials to connect to providers

In it's most simple form a generator doesn't need to carry credentials like the `Password` or `Fake`. But for more complex generators you want to talk to an API that generates a value for you, for example Vault/AWS (https://github.com/external-secrets/external-secrets/pull/1211) and many other Vault/Database, Vault/Kubernetes, Vault/PKI..., generate GitHub PAT tokens, ACR/ECR/GCR tokens etc. 

These more complex generators don't differ much from a SecretStore. The major difference is, that they provide access to a predefined set of keys that can be pulled (e.g. `spec.data` or `spec.dataFrom`).

There is - for sure - quite a lot of overlap and that leads to the question when would you implement a generator or a SecretStore for a given API backend.

This is all TBD; What were your expectations? IMHO it's a bit odd to have a SecretStore that provides access to a predefined set of keys (see use-cases above).

Reply:
For me the issue is that the ExternalSecrets manifest would start to carry out credentials information.. this means to impose controls there starts to become (alledgedly) more difficult (as in, a properly configured admission controller would need to be in place to avoid people with ExternalSecrets RBAC to point to other source of information.

Having the credentials inline with External Secrets also make a little bit shady how that would integrate with other setups (like ClusterSecretStores). AFAIK it would only work if and only if the secrets are available in the same namespace. Does this mean the generator feature is always namespaced?

Reply:
> ExternalSecrets manifest would start to carry out credentials information

only if the users goes with an inline generator. I'd not recommend it by any means but there are use-cases for that (creating TLS certs for instance).
Generally, users should use `sourceRef.generatorRef` in favour of inline generator.


Maybe, we can split it up: having a generator like this with auth information:
```yaml
apiVersion: generators.external-secrets.io/v1alpha1
kind: Foo
spec:
  auth:
    someAuthMethod: {} # ...
``` 

...and then make it configurable from the ExternalSecret so we can pass parameters
to the generator:
```yaml
apiVersion: external-secrets.io/v1beta1
kind: ExternalSecret
metadata:
  name: example
spec:
  # ...
  dataFrom:
    - generatorRef:
        apiVersion: generators.external-secrets.io/v1alpha1
        kind: Foo
        name: foo
        # add this:
        parameters:
          myparam: foobar
```
But that raises the question if we even need a generator apiGroup at all and not instead just make SecretStore configurable by passing a param `map[string]interface{}` to it?

> Does this mean the generator feature is always namespaced

No, people should use `sourceRef.generatorRef` to point to a (TBD) cluster-scoped generator.

Reply:
> No, people should use sourceRef.generatorRef to point to a (TBD) cluster-scoped generator.

Agree!

> only if the users goes with an inline generator. I'd not recommend it by any means but there are use-cases for that (creating TLS certs for instance).

Yeah, forgot about that. Inline does have a nice experience with `Password` generator, so I guess for now we could just have it in the docs to consider security boundaries when using inline generators ? OR even have the option to limit which generators are allowed to be rendered inline and which are not (not configurable options, just the code not allowing it to be rendered).

In any case I believe this is not a blocking point from what it's here :) Just further discussions we need to have before promoting it to something more stable.

Reply:
> But that raises the question if we even need a generator apiGroup at all and not instead just make SecretStore configurable by passing a param map[string]interface{} to it?

FTR: We've discussed that in the community meeting today. Key takeaway: there should be a semantical difference between a Generator (generates values) and a SecretStore (provides access to a remote API). Even though they have some overlap in functionality.

We'll drop support for inline generators for now (so right now no way of passing `map[string]interface{}` to a generator. We'll revisit the need for passing values to a Generator from an ExternalSecret. The community should decide where to go from here.